modules/gdscript/tests/scripts/analyzer/features/typed_array_usage.gd

```diff
@@ -0,0 +1,50 @@
+# https://github.com/godotengine/godot/pull/69248
+
+class A: pass
+class B extends A: pass
+
+func print_i(array: Array[int]) -> void: print(array)
+func print_a(array: Array[A]) -> void: print(array.size())
+func print_b(array: Array[B]) -> void: print(array.size())
+
+func print_def(array: Array[int] = [5]) -> void: print(array)
+
+const cm: Array[int] = [4]
+
+func test():
+	print_a([A.new()]) # 1
+	print_b([B.new()]) # 1
+	print_a([B.new()]) # 1
+
+	var a: Array[A] = [B.new()]
+	var b: Array[B] = [B.new()]
+	print_a(a) # 1
+	# print_b(a)
+	a = b
```

Comment:
Such assigment shouldn't be allowed. Arrays are reference types so if such assigment would be allowed then it should result in both `a` and `b` referencing the same array (the one `b` is initialized with). But `a` would still be typed as `Array[A]` and `b` would still be typed as `Array[B]`. Meaning the subsequent `a.push_back(A.new())`, although being typewise correct according to the type of `a` and the type of the argument, would be incorrect as the orignal array `a` is referencing is typed as `Array[B]` and shouldn't allow adding elements of a non-`B`-derived types like `A`.

Assigning `Array[Derived]` to a variable (non-`const`) of type `Array[Base]` shouldn't be allowed.

Reply:
Ok, just to be clear. This logic was not introduced by me. It was done properly before in case of assignment of an array of variants to an array of primitives - [here](https://github.com/godotengine/godot/blob/eba33c67b49905973ee7c81a709046415a41e205/core/variant/array.cpp#L227). And it was done improperly in other type mismatches with comments implying that the intention was to do the same - to copy contents.

I was just fixing it, not changing the previously intended behavior.

You saying that it should be changed to forbid such things?

I personally have no real objection to forbidding such assignment, it will make things more strict which generally I do prefer.

Reply:
I'd love for this to be clarified also. Are arrays meant to always be reference, or do they follow the "copy on write" policy? If we want them to be reference only, then I agree this needs to be stricter.

Which it wasn't originally! And possibly could just be another PR :)

Reply:
@vonagam To be clear as well: that's just my opinion. I'm not a GDScript maintainer or anything like that. So I'd wait for more opinions on that. But if it was already like that then I think potentially changing it to what I suggested is out of the scope of this PR anyway.

>I was just fixing it, not changing the previously intended behavior.

I'd say the question is whether it was indeed the intended behavior, whether it was thought over at all. And if it's not settled what the behavior should be then I'd say the tests added in this PR shouldn't include such cases. I think tests should be testing what is agreed to be the expected behavior. If there is no agreement on what the expected behavior is in some cases then I see no point in adding tests for such cases (these would just ensure things work like they currently do, which is not necessarily how they're supposed to work).

---

I think that assigment making a copy in case of typed arrays of non matching types is at least surprising. But I think for sure there should be a way to convert (by copying) a typed array of one type to a typed array of another type (like `Array[float]` to `Array[int]` or `Array[B]` -> `Array[A]`). However, currently it seems doing such conversion is possible only by that assignment (CMIIW). So in case it would be decided to be forbidden then some new way to convert typed arrays would need to be added.

---

When an assigment making a copy could make sense? When assigning a const to non-const. Like:
```gdscript
const const_array := [1, 2]
var non_const_array = const_array
```
Of course it could be copy on write, no need to copy right when the assigment is done.

---

All in all currently things are inconsistent, e.g. as mentioned assigning `Array[Derived]` to `Array[Base]` is allowed (and will create a new array) but the same case errors out when it's about passing it to a function:
```gdscript
func foo(a: Array[Base]) -> void:
	pass

func test() -> void:
	var b: Array[Derived] = ...
	foo(b) # Error!
```

---

Sorry for the late response and no real "what to do". To me it feels like we're lacking some kind of general GDScript design document, a guideline every contributor could look into. Seems like we're designing the language in a case by case manner when bug fixing, instead of the language being designed in the first place (then we could be fixing it according to the design). But I think it's a little too late to do anything about it now. Oh, well. :slightly_smiling_face:

Back to your PR: I haven't really looked into the code, just checked the tests and said about what catched my eye. Regarding the GDScript code I think e.g. @anvilfolk, @adamscott, or @rune-scape are more into it / are more familiar with it than I am, and hence probably will be better reviewers.
So yeah, feel free to don't do anything in this PR about the issue I've mentioned, it can be always tackled in another PR.

Reply:
Completely agree with what you said. I just did not question that `_assign` does two (currently three, but one is obviously wrong) different things - changes reference or changes contents.

I think `typed_assign` method should be about changing contents and `=` operator on a variable should be about changing reference.

Reply:
With new behavior it is no longer allowed. Resolved.